ibis/pandas/execution/strings.py

```diff
@@ -26,16 +26,50 @@ def execute_string_length_series(op, data, **kwargs):
     return data.str.len().astype('int32')
 
 
-@execute_node.register(
-    ops.Substring,
-    pd.Series,
-    (pd.Series,) + integer_types,
-    (pd.Series,) + integer_types
-)
-def execute_string_substring(op, data, start, length, **kwargs):
+@execute_node.register(ops.Substring, pd.Series, integer_types, integer_types)
+def execute_string_substring_int_int(op, data, start, length, **kwargs):
     return data.str[start:start + length]
 
 
+@execute_node.register(ops.Substring, pd.Series, pd.Series, integer_types)
+def execute_string_substring_series_int(op, data, start, length, **kwargs):
+    def iterate(value, it=start.values.flat):
+        begin = next(it)
+        if begin is not None and pd.isnull(begin):
+            return None
+        end = begin + length
+        return value[begin:end]
+    return data.map(iterate)
+
+
+@execute_node.register(ops.Substring, pd.Series, integer_types, pd.Series)
+def execute_string_substring_int_series(op, data, start, length, **kwargs):
+    end = start + length
+
+    def iterate(value, it=end.values.flat):
+        end = next(it)
+        if end is not None and pd.isnull(end):
+            return None
+        return value[start:end]
+    return data.map(iterate)
+
+
+@execute_node.register(ops.Substring, pd.Series, pd.Series, pd.Series)
+def execute_string_substring_series_series(op, data, start, length, **kwargs):
+    end = start + length
+
+    def iterate(value,
```

Comment:
This "general" case wouldn't be enough?

Reply:
I suppose we could farm out the other functions to this one.

Reply:
We definitely need to handle the integer, series and series, integer cases distinctly, but those two can be made to be instances of the general case.